File: 3.1_PermMissingElem.py
# returns the missing element of numeric array A

# Assumptions:
## 0 <= N <= 100,000
## the elements of A are all distinct
## 1<= A[i] <= N+1

def PermMissingElem(A):

	# exercise refers to N as the length of A, zero not included
	N = len(A)	

	# if set is missing 1 (including if empty),
	if(N==0) or not min(A)==1:

	# then 1 is the missing element
		return(1)

 	# if no skips exist in set (and it begins at 1),
	elif max(A)==N:

	# missing element is at end of set
		return(N+1) 

	# otherwise missing element is within set
	else:		

		# therefore, first sort the set
		A = sorted(A)

		# then crawl along array and test each element
		for n in range(0, N):

			# A[n] will correspond to n+1 until skipped element
			if A[n]>n+1:

				# then, retun n+1
				return(n+1)

# Console Testing:
print(["Empty list - []; expect 1: "]+[PermMissingElem([])])
print(["Single element - [1]; expect 2: "]+[PermMissingElem([1])])
print(["Missing first - [3,2,5,4]; expect 1: "]+[PermMissingElem([3,2,5,4])])
print(["Missing last - [1,2,4,3]; expect 5: "]+[PermMissingElem([1,2,4,3])])
print(["Double - [1,3]; expect 2: "]+[PermMissingElem([1,3])])	# got None expected 2
print(["Practice - [2,3,1,5]; expect 4: "]+[PermMissingElem([2,3,1,5])])

# Codility testing: 
## https://app.codility.com/programmers/lessons/3-time-complexity/perm_missing_elem/
## correctness = 100%
## performance = 100%

File: 3.1_PermMissingElem.R
# returns the missing element of numeric array A

# Assumptions:
## 0 <= N <= 100,000
## the elements of A are all distinct
## 1<= A[i] <= N+1

PermMissingElem <- function(A){
	A <- as.numeric(unlist(A))	# convert to numeric vector
	N <- length(A)			# define N as length of A, as given
	if(N==0 | min(A)!=1){		# if set is missing 1 (including if empty),
		return(1)			# then missinge element is 1
	}else if(max(A)==N){		# if no skips exist in set (assume it begins at 1),
		return(max(A)+1)		# then missing integer is at end of (sorted) set
	}else{				# otherwise missing element is between max and min, so:
		A <- sort(A)			# 1) sort the set
		for(n in 1:N){			# 2) crawl through the set
			if(A[n]>n){		# 3) A[n] > n when crawl skips an integer
				return(n)		# then n = missing integer
			}
		}
	}
}

# Console Testing:
print(c("Empty list - list()", PermMissingElem(list())))
print(c("Single element - list(3)", PermMissingElem(list(3))))
print(c("Single element - list(1)", PermMissingElem(list(1))))
print(c("Missing first - list(3,2,5,4)", PermMissingElem(list(3,2,5,4))))
print(c("Missing last - list(1,2,4,3)", PermMissingElem(list(1,2,4,3))))
print(c("Double - list(1,3)", PermMissingElem(list(1,3))))
print(c("Practice - list(2,3,1,5)", PermMissingElem(list(2,3,1,5))))

# Codility testing: 
## https://app.codility.com/programmers/lessons/3-time-complexity/perm_missing_elem/
## R language not scored
## Difficulty: Painless
